client/src/main/java/org/eclipse/hono/client/impl/HonoConnectionImpl.java

```diff
@@ -597,14 +597,19 @@ private void failConnectionAttempt(final Throwable connectionFailureCause, final
      *
      * @param link The link to close. If {@code null}, the given handler is invoked immediately.
      * @param closeHandler The handler to notify once the link has been closed.
-     * @throws NullPointerException if context or close handler are {@code null}.
+     * @throws NullPointerException if close handler is {@code null}.
      */
     @Override
     public void closeAndFree(
             final ProtonLink<?> link,
             final Handler<Void> closeHandler) {
 
-        HonoProtonHelper.closeAndFree(context, link, closeHandler);
+        if (context == null) {
+            // this means that the connection to the peer is not established (yet)
+            closeHandler.handle(null);
+        } else {
+            HonoProtonHelper.closeAndFree(context, link, closeHandler);
```

Comment:
why don't we simply fix this in `HonoProtonHelper.closeAndFree()`?

Reply:
I would rather require `HonoProtonHelper.closeAndFree()` to get a non-null Vert.x `context`, because I think that `context` being non-null should be the standard case.

The case of HonoConnection having a null `context` in case the connection is not established (yet) is not really intuitive to me and therefore should not influence how the general `HonoProtonHelper` is implemented, IMHO.